File: python/veval/_test_sdk.py
from __future__ import annotations
import uuid
from collections import deque
from datetime import datetime, timezone
from typing import Any, Callable, Awaitable, Optional, TypeVar

from ._options import VevalOptions
from ._context import VevalExecutionContext
from ._tracing import TraceData, SnapshotData, SnapshotDiff, compare_snapshots, ReplayOptions, ReplayResult
from ._assertions import ITraceAssertion
from ._scenarios import ScenarioItem, ScenarioRunResult, ItemRunResult
from ._http_client import VevalHttpClient
from ._sdk import VevalSdk, _flatten_steps

T = TypeVar("T")


class VevalTestSdk(VevalSdk):
    """
    Test double for VevalSdk. Guarantees no live LLM calls.
    Always reports to the dashboard so runs count toward quota.

    Usage:
        sdk = VevalTestSdk(options).with_replay(trace)
        # inject into your service, call normally.
    """

    def __init__(self, options: VevalOptions):
        super().__init__(options)
        self._replay_trace: Optional[TraceData] = None
        self._last_status: Optional[str] = None
        self._last_error: Optional[str] = None
        self._judge_mocks: dict[str, deque] = {}

    @property
    def last_status(self) -> Optional[str]:
        return self._last_status

    @property
    def last_error(self) -> Optional[str]:
        return self._last_error

    def with_replay(self, trace: TraceData) -> VevalTestSdk:
        self._replay_trace = trace
        return self

    def with_judge_mock(
        self,
        criteria: str,
        passed: bool,
        score: float = 1.0,
        reasoning: str = "",
    ) -> VevalTestSdk:
        """
        Registers a canned judge result for the given criteria string, so scenarios/replays
        using TraceAssert.judge stay deterministic and free in CI. Without a matching mock,
        judge_async raises rather than silently making a real, billed LLM call.
        """
        self._judge_mocks.setdefault(criteria, deque()).append(
            {"score": score, "passed": passed, "reasoning": reasoning}
        )
        return self

    async def judge_async(
        self,
        criteria: str,
        ctx: VevalExecutionContext,
        model: Optional[str] = None,
    ) -> dict:
        queue = self._judge_mocks.get(criteria)
        if queue:
            return queue.popleft()
        raise RuntimeError(
            f"Replay mode: no judge mock for criteria '{criteria}'. "
            "Call with_judge_mock(...) before running a scenario/replay that uses "
            "TraceAssert.judge — this would have made a real, billed LLM call."
        )

    async def run_async(
        self,
        agent_name: str,
        callback: Callable[[VevalExecutionContext], Awaitable[T]],
        input: Any = None,
    ) -> T:
        ctx = VevalExecutionContext(f"tr_{uuid.uuid4().hex}", input)
        if self._replay_trace:
            ctx.load_mock_outputs(self._replay_trace)

        started_at = datetime.now(timezone.utc)
        output: Any = None
        status = "success"
        error: Optional[str] = None

        try:
            output = await callback(ctx)
            self._last_status = "success"
            self._last_error = None
        except Exception as ex:
            status = "error"
            error = str(ex)
            self._last_status = "error"
            self._last_error = error
            raise
        finally:
            completed_at = datetime.now(timezone.utc)
            extra_meta: dict[str, Any] = {"replay": True}
            if self._replay_trace:
                extra_meta["source_trace_id"] = self._replay_trace.trace_id
            payload = self._build_payload(
                ctx.trace_id, agent_name, ctx,
                input, output, status, error,
                started_at, completed_at,
                extra_meta=extra_meta,
            )
            await self._client.send_trace_async(payload)

        return output  # type: ignore[return-value]

    async def get_trace_async(self, trace_id: str) -> Optional[TraceData]:
        if self._replay_trace and self._replay_trace.trace_id == trace_id:
            return self._replay_trace
        return None

    async def load_snapshot_async(self, trace_id: str) -> Optional[SnapshotData]:
        trace = await self.get_trace_async(trace_id)
        return SnapshotData.from_trace(trace) if trace else None

    async def run_scenario_async(
        self,
        scenario_name: str,
        agent: Callable[[VevalExecutionContext], Awaitable[T]],
        scenario_assertions: list[ITraceAssertion],
        items: Optional[list[ScenarioItem]] = None,
    ) -> ScenarioRunResult:
        resolved_items = items or []
        scenario_result = ScenarioRunResult()

        for item in resolved_items:
            effective_assertions = scenario_assertions + list(item.assertions)
            item_result = ItemRunResult(item=item)

            if item.trace_id:
                trace = await self.get_trace_async(item.trace_id)
                if trace is None:
                    item_result.failures.append(
                        f"trace '{item.trace_id}' not found — call with_replay(trace) before run_scenario_async"
                    )
                else:
                    replay_result = await self.replay_async(
                        trace, agent,
                        ReplayOptions(mock_llm_responses=True, assertions=effective_assertions),
                    )
                    item_result.failures.extend(replay_result.failures)
                    item_result.context = replay_result.replayed_context

                    if replay_result.replayed_context:
                        replay_ctx = replay_result.replayed_context
                        payload = self._build_payload(
                            replay_ctx.trace_id, scenario_name, replay_ctx,
                            trace.input, replay_result.output,
                            replay_result.status, replay_result.error,
                            replay_result.started_at, replay_result.completed_at,
                            extra_meta={"replay": True, "source_trace_id": item.trace_id},
                        )
                        await self._client.send_trace_async(payload)

            elif item.input is not None:
                ctx = VevalExecutionContext(f"tr_{uuid.uuid4().hex}", item.input)
                try:
                    await agent(ctx)
                except Exception as ex:
                    item_result.failures.append(f"Agent threw exception: {ex}")

                for assertion in effective_assertions:
                    failure = await assertion.evaluate_async(ctx)
                    if failure:
                        item_result.failures.append(failure)
                item_result.context = ctx
            else:
                item_result.failures.append("ScenarioItem must have either trace_id or input")

            scenario_result.results.append(item_result)

        await self._client.post_scenario_run_async(scenario_name, {
            "passed": scenario_result.passed,
            "pass_count": scenario_result.pass_count,
            "fail_count": scenario_result.fail_count,
            "results": [
                {
                    "name": r.item.name or r.item.trace_id or "item",
                    "passed": r.passed,
                    "failures": r.failures,
                }
                for r in scenario_result.results
            ],
        })

        return scenario_result

File: node/src/test-sdk.ts
import { VevalOptions } from "./options";
import { VevalSdk } from "./sdk";
import { VevalExecutionContext } from "./context";
import { TraceData, ReplayOptions } from "./tracing";
import { SnapshotData } from "./snapshots";
import { ITraceAssertion, JudgeResult } from "./assertions";
import { ScenarioItem, ScenarioRunResult, ItemRunResult } from "./scenarios";

export class VevalTestSdk extends VevalSdk {
  private _replayTrace: TraceData | null = null;
  private _lastStatus: string | null = null;
  private _lastError: string | null = null;
  private _judgeMocks: Map<string, JudgeResult[]> = new Map();
  private _snapshots: Map<string, SnapshotData> = new Map();

  constructor(options: VevalOptions) {
    super(options);
  }

  get lastStatus(): string | null {
    return this._lastStatus;
  }

  get lastError(): string | null {
    return this._lastError;
  }

  withReplay(trace: TraceData): this {
    this._replayTrace = trace;
    return this;
  }

  /**
   * Registers a canned judge result for the given criteria string, so scenarios/replays
   * using TraceAssert.judge stay deterministic and free in CI. Without a matching mock,
   * judgeAsync throws rather than silently making a real, billed LLM call.
   */
  withJudgeMock(criteria: string, passed: boolean, score = 1.0, reasoning = ""): this {
    const queue = this._judgeMocks.get(criteria) ?? [];
    queue.push({ score, passed, reasoning });
    this._judgeMocks.set(criteria, queue);
    return this;
  }

  override async judgeAsync(criteria: string): Promise<JudgeResult> {
    const queue = this._judgeMocks.get(criteria);
    const next = queue?.shift();
    if (next) return next;
    throw new Error(
      `Replay mode: no judge mock for criteria '${criteria}'. ` +
        "Call withJudgeMock(...) before running a scenario/replay that uses " +
        "TraceAssert.judge — this would have made a real, billed LLM call."
    );
  }

  override async runAsync<T>(
    agentName: string,
    callback: (ctx: VevalExecutionContext) => Promise<T>,
    input?: unknown
  ): Promise<T> {
    const traceId = `tr_${crypto.randomUUID().replace(/-/g, "")}`;
    const ctx = new VevalExecutionContext(traceId, input ?? null);
    if (this._replayTrace) {
      ctx.loadMockOutputs(this._replayTrace);
    }

    const startedAt = new Date();
    let output: unknown = null;
    let status = "success";
    let error: string | null = null;

    try {
      output = await callback(ctx);
      this._lastStatus = "success";
      this._lastError = null;
    } catch (err) {
      status = "error";
      error = err instanceof Error ? err.message : String(err);
      this._lastStatus = "error";
      this._lastError = error;

      const completedAt = new Date();
      const extraMeta: Record<string, unknown> = { replay: true };
      if (this._replayTrace) extraMeta["source_trace_id"] = this._replayTrace.trace_id;
      const payload = VevalSdk.buildPayload(
        traceId, agentName, this.opts.projectId, ctx,
        input ?? null, null, status, error, startedAt, completedAt, extraMeta
      );
      await this.http.sendTraceAsync(payload);
      throw err;
    }

    const completedAt = new Date();
    const extraMeta: Record<string, unknown> = { replay: true };
    if (this._replayTrace) extraMeta["source_trace_id"] = this._replayTrace.trace_id;
    const payload = VevalSdk.buildPayload(
      traceId, agentName, this.opts.projectId, ctx,
      input ?? null, output, status, error, startedAt, completedAt, extraMeta
    );
    await this.http.sendTraceAsync(payload);
    return output as T;
  }

  /**
   * Registers a baseline locally, so getSnapshotAsync / TraceAssert.matchesSnapshot work offline in CI.
   * Names without a local baseline are loaded from the server.
   */
  withSnapshot(snapshotName: string, snapshot: SnapshotData): this {
    this._snapshots.set(snapshotName, snapshot);
    return this;
  }

  override async getSnapshotAsync(snapshotName: string): Promise<SnapshotData | null> {
    return this._snapshots.get(snapshotName) ?? super.getSnapshotAsync(snapshotName);
  }

  override async getTraceAsync(traceId: string): Promise<TraceData | null> {
    if (this._replayTrace?.trace_id === traceId) return this._replayTrace;
    return null;
  }

  override async runScenarioAsync<T>(
    scenarioName: string,
    agent: (ctx: VevalExecutionContext) => Promise<T>,
    scenarioAssertions: ITraceAssertion[],
    items?: ScenarioItem[]
  ): Promise<ScenarioRunResult> {
    const resolvedItems = items ?? [];
    const results: ItemRunResult[] = [];

    for (const item of resolvedItems) {
      const effectiveAssertions = [...scenarioAssertions, ...(item.assertions ?? [])];
      const itemResult: ItemRunResult = { item, passed: true, failures: [] };

      if (item.trace_id) {
        const trace = await this.getTraceAsync(item.trace_id);
        if (!trace) {
          itemResult.failures.push(
            `trace '${item.trace_id}' not found — call withReplay(trace) before runScenarioAsync`
          );
          itemResult.passed = false;
        } else {
          const replayResult = await this.replayAsync(trace, agent, {
            mock_llm_responses: true,
            assertions: effectiveAssertions,
          } as ReplayOptions);
          itemResult.failures.push(...replayResult.failures);
          itemResult.passed = replayResult.passed;
          itemResult.context = replayResult.replayed_context ?? undefined;

          if (replayResult.replayed_context) {
            const replayCtx = replayResult.replayed_context;
            const payload = VevalSdk.buildPayload(
              replayCtx.traceId, scenarioName, this.opts.projectId, replayCtx,
              trace.input, replayResult.output, replayResult.status, replayResult.error,
              new Date(replayResult.started_at), new Date(replayResult.completed_at),
              { replay: true, source_trace_id: item.trace_id }
            );
            await this.http.sendTraceAsync(payload);
          }
        }
      } else if (item.input !== undefined) {
        const ctx = new VevalExecutionContext(`tr_${crypto.randomUUID().replace(/-/g, "")}`, item.input);
        try {
          await agent(ctx);
        } catch (err) {
          const msg = err instanceof Error ? err.message : String(err);
          itemResult.failures.push(`Agent threw exception: ${msg}`);
          itemResult.passed = false;
        }
        for (const assertion of effectiveAssertions) {
          const failure = await assertion.evaluate(ctx);
          if (failure) {
            itemResult.failures.push(failure);
            itemResult.passed = false;
          }
        }
        itemResult.context = ctx;
      } else {
        itemResult.failures.push("ScenarioItem must have either trace_id or input");
        itemResult.passed = false;
      }

      results.push(itemResult);
    }

    const passCount = results.filter((r) => r.passed).length;
    const failCount = results.length - passCount;
    const scenarioResult: ScenarioRunResult = {
      passed: failCount === 0,
      pass_count: passCount,
      fail_count: failCount,
      results,
    };

    await this.http.postScenarioRunAsync(scenarioName, {
      passed: scenarioResult.passed,
      pass_count: scenarioResult.pass_count,
      fail_count: scenarioResult.fail_count,
      results: results.map((r) => ({
        name: r.item.name ?? r.item.trace_id ?? "item",
        passed: r.passed,
        failures: r.failures,
        judgments: r.context?.judgments ?? [],
      })),
    });

    return scenarioResult;
  }
}
